### ISSUE TITLE
(Critical) eval exposed in Source 3

### ISSUE THREAD

**Issue by openorclose** [source-academy/js-slang#81]
```javascript
pair.constructor("return window;")();
```
Basically any code can be run.

**Comment by martin-henz** [comment]
Looks like Source §3 is quite permissive.

But I don't understand what's "critical" about this. The "." operator is not in Source §3, so all of these things (including +=, var) are undocumented features. They should not be in Source §3, but they currently don't lead to an error.

Or are there any security issues with this?

**Comment by openorclose** [comment]
There's no extra security issues, since one can run JavaScript from the console anyway.

The main difference between eval and the rest is that the rest of the undocumented features fail silently without providing additional functionality, while eval allows any valid JavaScript code to run.

Students may think that whatever can be run in the playground is valid Source, and use these extra undocumented features in doing their missions.

**Comment by martin-henz** [comment]
I agree that some students might think that way. However, we make it quite clear that they need to use Source.

In any case, let's try to fix these issues. Good to be consistent. No need to be open to confusion in this way.

Thanks for pointing out the issue.

**Comment by openorclose** [comment]
From grader:

1. Receives a student program string, and an array of grader program strings

2. For each grader program,

> - Concatenate the student program and grader program into a single combined program
> - Evaluates the single combined program in the js-slang interpreter

Is this how students' programs are being evaluated? 

If it is, then yes, then the existence of eval in Source is a critical security issue. It will be possible to execute potentially malicious code on if the grader runs on a server. 

If the grader is run locally and not on a server, at the very least eval will allow code to change how the autograder works.

**Comment by ningyuansg** [comment]
Yes to all of the above, save that the grader runs on AWS lambda.

**Comment by martin-henz** [comment]
Yes, I agree that this is a security issue. Thanks for pointing this out. At the moment, for our missions, we are eye-balling the student's programs, but for our Practical Assessment in November, such hacking attempt might fool our semi-automated process. So yes, we need to fix this! Thanks a lot @openorclose for pointing this out!

**Comment by martin-henz** [comment]
This is still not fixed:
pair.constructor("return window;")().alert(1);
